File: scala/leetcode_scala/src/main/scala/leetcode/editor/cn/[160]相交链表.scala
package leetcode.editor.cn
object IntersectionOfTwoLinkedLists {
//leetcode submit region begin(Prohibit modification and deletion)
  /** Definition for singly-linked list.
    * class ListNode(var _x: Int = 0) {
    *   var next: ListNode = null
    *   var x: Int = _x
    * }
    */

  object Solution {
    def getIntersectionNode(headA: ListNode, headB: ListNode): ListNode = {
      @annotation.tailrec
      def helper(p1: ListNode, p2: ListNode): ListNode = {
        if (p1 == p2) p1
        else {
          val nextP1 = if (p1 == null) headB else p1.next
          val nextP2 = if (p2 == null) headA else p2.next
          helper(nextP1, nextP2)
        }
      }
      if (headA == null || headB == null) null
      else helper(headA, headB)
    }
  }
//leetcode submit region end(Prohibit modification and deletion)

}

File: cpp/leetcodecn/[160]相交链表.cpp
#include <vector>
#include "util.h"
#include <map>
#include <string>

using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        auto p1 = headA;
        auto p2 = headB;
        while (p1 != p2) {
            if (p1 == nullptr)
                p1 = headA;
            else
                p1 = p1->next;

            if (p2 == nullptr)
                p2 = headB;
            else
                p2 = p2->next;
        }

        return p1;
    }
};
//leetcode submit region end(Prohibit modification and deletion)


int main(int argv, char *argc[]) {
}